src/pyodide-py/pyodide/simple_web_loop.py

```diff
@@ -0,0 +1,203 @@
+        because we only have one thread so blocking would stall the browser event loop
+        and prevent anything from ever happening.
+        """
+        pass
+
+    def run_until_complete(self, future: Awaitable):
+        """ Since we cannot block, we just ensure that the future is scheduled. """
+        return asyncio.ensure_future(future)
+
+    #
+    # Scheduling methods: use browser.setTimeout to schedule tasks on the browser event loop.
+    #
+
+    def call_soon(self, callback: Callable, *args, context: contextvars.Context = None):
+        """
+        Schedule the callback callback to be called with args arguments at the next iteration of the event loop.
+        """
+        delay = 0
+        return self.call_later(delay, callback, *args, context=context)
+
+    def call_soon_threadsafe(
+        callback: Callable, *args, context: contextvars.Context = None
+    ):
+        """
+        A thread-safe variant of call_soon().
+
+        Note this function is different from the standard asyncio loop implementation, it is current exactly the same as call_soon
+        """
+        return self.call_soon(callback, *args, context=context)
+
+    def call_later(
+        self,
+        delay: float,
+        callback: Callable,
+        *args,
+        context: contextvars.Context = None
+    ):
+        """
+        Schedule callback to be called after the given delay number of seconds (can be either an int or a float).
+        """
+        from js import setTimeout
+
+        if delay < 0:
+            raise Exception("Can't schedule in the past")
+        h = asyncio.Handle(callback, args, self, context=context)
+        setTimeout(h._run, delay * 1000)
+        return h
+
+    def call_at(
+        self,
+        when: float,
+        callback: Callable,
+        *args,
+        context: contextvars.Context = None
+    ):
+        """
+        Schedule callback to be called at the given absolute timestamp when (an int or a float), using the same time reference as loop.time().
+        """
+        cur_time = self.time()
+        delay = when - cur_time
+        return self.call_later(delay, callback, *args, context=context)
+
+    #
+    # The remaining methods are copied directly from BaseEventLoop
+    #
+
+    def time(self):
+        """Return the time according to the event loop's clock.
+
+        This is a float expressed in seconds since an epoch, but the
+        epoch, precision, accuracy and drift are unspecified and may
+        differ per event loop.
+
+        Copied from BaseEventLoop.time
+        """
+        return time.monotonic()
+
+    def create_future(self):
+        """Create a Future object attached to the loop.
+
+        Copied from BaseEventLoop.create_future
+        """
+        return futures.Future(loop=self)
+
+    def create_task(self, coro, *, name=None):
+        """Schedule a coroutine object.
+
+        Return a task object.
+
+        Copied from BaseEventLoop.create_task
```

Comment:
> BaseEventLoop should not be subclassed by third-party code; the internal interface is not stable.

It's a shame that we have to re-define everything.

Reply:
Yeah I thought about subclassing `BaseEventLoop` but it has a _lot_ of logic in it and most of it should never be invoked by our event loop, and also they say not to do it. So it's better to just copy these five methods, they're all really short anyways.